### PR TITLE
Refactor CMake code for dictionary regeneration

### PR DESCRIPTION
**Pull request by Teemperor**
Just to test in jenkins

### CODE REVIEW

**Comment by Teemperor** [comment]
Still needs a lot of formatting, will do when I have time.

**Comment by martinmine** [comment]
@phsft-bot build and stop being lazy

**Comment by Teemperor** [comment]
@phsft-bot build

File: cmake/modules/RootNewMacros.cmake [hunk 1]
```diff
@@ -827,13 +827,70 @@ function(ROOT_INSTALL_HEADERS)
...
+  # Check if we have any unparsed arguments
+  if(ARG_UNPARSED_ARGUMENTS)
+    message(AUTHOR_WARNING "Unparsed arguments for ROOT_STANDARD_LIBRARY_PACKAGE: ${ARG_UNPARSED_ARGUMENTS}")
+  endif()
+
+  if (NOT ARG_HEADERS)
+    set(ARG_HEADERS "*.h")
+  endif()
+  if (NOT ARG_SOURCES AND NOT ARG_HEADER_ONLY)
+    set(ARG_SOURCES "*.cxx")
```
Comment (amadio) [inline cmake/modules/RootNewMacros.cmake]: Should this be "src/*.cxx"?

Comment (Teemperor) [inline cmake/modules/RootNewMacros.cmake]: Yeah, but I didn't want to change ROOT_STANDARD_LIBRARY_PACKAGE behavior (right now it behaves exactly the same way as before as long as you don't use the new arguments). Maybe we should indeed use a new function. Let's discuss this on monday.

**Comment by Teemperor** [comment]
@phsft-bot build

jenkins crashed during checkout...

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by vgvassilev** [comment]
Can you rename the commit message. Looks good overall, let me know if I should test it on massively parallel machine.

File: cmake/modules/RootNewMacros.cmake [hunk 2]
```diff
@@ -820,13 +820,77 @@ function(ROOT_INSTALL_HEADERS)
 endfunction()
 
 #---------------------------------------------------------------------------------------------------
-#---ROOT_STANDARD_LIBRARY_PACKAGE(libname DEPENDENCIES lib1 lib2)
+#---ROOT_STANDARD_LIBRARY_PACKAGE(libname
+#                                 [NO_INSTALL_HEADERS]         : don't install headers for this package
+#                                 [STAGE1]                     : use rootcling_stage1 for generating
+#                                 HEADER header1 header2       : if not specified, globbing for *.h is used)
```
Comment (peremato) [inline cmake/modules/RootNewMacros.cmake]: The parameter is HEADERS and not HEADER

File: cmake/modules/RootNewMacros.cmake [hunk 3]
```diff
@@ -820,13 +820,77 @@ function(ROOT_INSTALL_HEADERS)
...
+#                                 DICTIONARY_OPTIONS option    : options passed to rootcling
+#                                 INSTALL_OPTIONS option       : options passed to install headers
+#                                )
 #---------------------------------------------------------------------------------------------------
 function(ROOT_STANDARD_LIBRARY_PACKAGE libname)
-  CMAKE_PARSE_ARGUMENTS(ARG "" "" "DEPENDENCIES;DICTIONARY_OPTIONS" ${ARGN})
-  ROOT_GENERATE_DICTIONARY(G__${libname} *.h MODULE ${libname} LINKDEF LinkDef.h OPTIONS ${ARG_DICTIONARY_OPTIONS})
-  ROOT_LINKER_LIBRARY(${libname} *.cxx G__${libname}.cxx DEPENDENCIES ${ARG_DEPENDENCIES})
-  ROOT_INSTALL_HEADERS()
+  set(options OPTIONAL NO_INSTALL_HEADERS STAGE1 NO_HEADERS NO_SOURCES OBJECT_LIBRARY)
```
Comment (peremato) [inline cmake/modules/RootNewMacros.cmake]: Is OPTIONAL used anywhere?

File: geom/geocad/CMakeLists.txt [hunk 4]
```diff
@@ -8,8 +8,5 @@ include_directories(${OCC_INCLUDE_DIR})
 ROOT_ADD_CXX_FLAG(CMAKE_CXX_FLAGS -Wno-overloaded-virtual)
 ROOT_ADD_CXX_FLAG(CMAKE_CXX_FLAGS -Wno-ignored-qualifiers)
 
-ROOT_GENERATE_DICTIONARY(G__GeoCad *.h MODULE GeoCad LINKDEF LinkDef.h DEPENDENCIES Geom)
 
-ROOT_LINKER_LIBRARY(GeoCad *.cxx G__GeoCad.cxx LIBRARIES Core ${OCC_LIBRARIES} DEPENDENCIES Geom)
-
-ROOT_INSTALL_HEADERS()
+ROOT_STANDARD_LIBRARY_PACKAGE(GeoCad HEADERS *.h DEPENDENCIES Geom LIBRARIES ${OCC_LIBRARIES})
```
Comment (peremato) [inline geom/geocad/CMakeLists.txt]: probably HEADERS *.h is not needed

File: graf2d/graf/CMakeLists.txt [hunk 5]
```diff
@@ -3,11 +3,7 @@
 # @author Pere Mato, CERN
 ############################################################################
 
-ROOT_GENERATE_DICTIONARY(G__Graf MODULE Graf *.h LINKDEF LinkDef.h OPTIONS "-writeEmptyRootPCM" DEPENDENCIES Hist Matrix MathCore RIO BUILTINS FREETYPE)
-
-
 include_directories(${FREETYPE_INCLUDE_DIRS})
 
-ROOT_LINKER_LIBRARY(Graf *.cxx G__Graf.cxx LIBRARIES ${FREETYPE_LIBRARIES} ${ZLIB_LIBRARIES} mathtext DEPENDENCIES Hist Matrix MathCore RIO BUILTINS FREETYPE)
+ROOT_STANDARD_LIBRARY_PACKAGE(Graf HEADERS *.h DICTIONARY_OPTIONS "-writeEmptyRootPCM" LIBRARIES ${FREETYPE_LIBRARIES} ${ZLIB_LIBRARIES} mathtext DEPENDENCIES Hist Matrix MathCore RIO BUILTINS FREETYPE)
```
Comment (peremato) [inline graf2d/graf/CMakeLists.txt]: idem HEADERS *.h is probably not needed (default)

File: graf3d/gl/CMakeLists.txt [hunk 6]
```diff
@@ -37,8 +37,4 @@ endif()
 set_source_files_properties(src/TGLFontManager.cxx PROPERTIES COMPILE_FLAGS "${FTGL_CFLAGS}")
 set_source_files_properties(src/TGLText.cxx PROPERTIES COMPILE_FLAGS "${FTGL_CFLAGS}")
 
-ROOT_GENERATE_DICTIONARY(G__RGL ${headers} MODULE RGL LINKDEF LinkDef.h OPTIONS "-writeEmptyRootPCM" DEPENDENCIES Hist Gui Ged)
-
-ROOT_LINKER_LIBRARY(RGL ${sources} G__RGL.cxx  LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} DEPENDENCIES Hist Gui Ged)
-
-ROOT_INSTALL_HEADERS(${installoptions})
+ROOT_STANDARD_LIBRARY_PACKAGE(RGL HEADERS ${headers} SOURCES ${sources} DICTIONARY_OPTIONS "-writeEmptyRootPCM" LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} DEPENDENCIES Hist Gui Ged INSTALL_OPTIONS ${installoptions})
```
Comment (peremato) [inline graf3d/gl/CMakeLists.txt]: General comment. Some of these lines are too long. In some cases you have formatted nicely like:
```
ROOT_STANDARD_LIBRARY_PACKAGE(RGL HEADERS ${headers} 
                                  SOURCES ${sources} 
                                  DICTIONARY_OPTIONS "-writeEmptyRootPCM" 
                                  LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} 
                                  DEPENDENCIES Hist Gui Ged 
                                  INSTALL_OPTIONS ${installoptions})
```

**Comment by amadio** [comment]
@peremato We will test everything on several places before merging. The previous changes were tested by me on my laptop, on a server with 36 cores, and on a KNL machine. I used 6, 70, and 250 threads, respectively, and had no failures. The same testing will happen for this PR to ensure that nothing breaks. Then we merge and check that no failures show up on the nightlies and incrementals.

**Comment by peremato** [comment]
@amadio what I wanted to say is that it is not sufficient to run all the tests, which are run in the build directory and I assume will be done correctly, we need to check that the installation (make install) are complete and no header has been forgotten in the process. The new generic macro deals with the header installation and this is why we need to check carefully.

**Comment by vgvassilev** [comment]
Probably one easy way to assert that is to diff the modulemap files before and after this patch. They work with the headers to be installed, too.

**Comment by amadio** [comment]
We could also create a diff of a list of all files installed.

**Comment by vgvassilev** [comment]
ping...

**Comment by peremato** [comment]
Ping what? I reviewed 8 days ago.

**Comment by vgvassilev** [comment]
@peremato, it was a ping for @Teemperor.

**Comment by Teemperor** [comment]
Updated, fixed a typo in one of the windows-only CMake files, fixed a few more differences between what headers are installed and formatted everything. Can someone give it a test install to double-check the headers?

**Comment by amadio** [comment]
I created a tarball with the installed files from the master branch and this branch and I see no difference when I apply a diff on the list of installed files. @vgvassilev Feel free to merge.

**Pull request merged**